### DISCUSSION TITLE
How to eager load a list of ids that are not defined as a relation

### DISCUSSION THREAD

**Discussion by pc386** [rebing/graphql-laravel#927] [Q&A]
I have a table with a column called image_ids. It is of type json, an array of ids. The column is not configured as a relation. However it is a relation to the table images. So each of the image_ids exist in the table images. How can I eager load those so I don't have to query each when requesting it? Currently I have a custom resolve for the images field in my Type. Like this:

```
public function resolveImagesField($root): array
    {
        $images = [];
        if (!empty($root['image_ids'])) {
            $imageService = app()->make(ImageServiceInterface::class);
            foreach ($root['image_ids'] as $imageId) {
                $images[] = $imageService->getById($imageId)->getUrl();
            }
        }

        return $images;
    }
```
As you can see for each id in the image_ids, I query the image. 
However this is less than optimal since it won't be eager loaded and the performance tanks.

**Comment by mfn** [comment]
Doesn't your `ImageServiceInterface` have a `getByIds`?

But that would only solve it for one element, I assume you return multiple "things" which each have their `image_ids`, right?

**Reply by pc386** [reply]
That's an improvment already, but yes there are multiple things with image_ids, and I want those all to be eager loaded

**Reply by mfn** [reply]
Then you probably need the [dataloader approach](https://www.google.com/search?hl=en&q=graphql%20dataloader), which is quite a popular solution for GraphQL [and has PHP implementations too](https://www.google.com/search?hl=en&q=graphql%20dataloader%20php).

High level, in your resolver:
- you store the `image_ids` in your "image data loader" -> you just collect them there
- then you return a `new \GraphQL\Deferred()`
  - where you passed a closure to the `Deferred` which
    - _also_ has access to the "image data loader"
    - also pass thru the `image_ids` you collected before
    - but in here you "resolve" them, e.g. **lazy load** them

Couple of things to note:
- `\GraphQL\Deferred` is from the underlying `webonyx/graphql-php` library and is transparently supported
- I know you said "eager loading" and I responded with "lazy loading"
  The trick here is:
  - the `Deferred` is called **after all** `image_ids` have been collected
  - so when you resolve your _actual first_ `image_ids` for a specific item, you do in fact _load all of them at once_ (in your most efficient way; but also could just be `->whereIn(…)`)
  - but for any given resolver you _only_ return the `image_ids` you collected there

The "image data loader" will only load **all** your `image_ids` across all items once (=cache them internally) but return only the once requested by each particulars item resolver.

Hope I could explain it well enough. See also https://webonyx.github.io/graphql-php/data-fetching/#solving-n1-problem
